blazar/plugins/oshosts/host_plugin.py

```diff
@@ -1005,21 +1006,36 @@ def _poll_resource_failures(self):
         :return: a list of failed hosts, a list of recovered hosts.
         """
         hosts = db_api.host_get_all_by_filters({})
-        reservable_hosts = [h for h in hosts if h['reservable'] is True]
-        unreservable_hosts = [h for h in hosts if h['reservable'] is False]
+        reservable_ironic_hosts = [h for h in hosts if h['reservable'] is True and h['hypervisor_type'] == 'ironic']
+        unreservable_irinic_hosts = [h for h in hosts if h['reservable'] is False and h['hypervisor_type'] == 'ironic']
+        reservable_nova_hosts = [h for h in hosts if h['reservable'] is True and h['hypervisor_type'] != 'ironic']
+        unreservable_nova_hosts = [h for h in hosts if h['reservable'] is False and h['hypervisor_type'] != 'ironic']
+
+        failed_hosts = []
+        recovered_hosts = []
 
         try:
+            ironic_client = ironic.BlazarIronicClient()
+            bms = ironic_client.ironic.node.list()
+
+            failed_bm_ids = [bm.uuid for bm in bms
+                             if bm.maintenance or bm.power_state in ['error', None] or bm.provision_state not in ['active', 'available', None]]
+            failed_hosts = failed_hosts + [host for host in reservable_ironic_hosts
+                                           if host['hypervisor_hostname'] in failed_bm_ids]
+            recovered_hosts = recovered_hosts + [host for host in unreservable_irinic_hosts
+                                                 if host['hypervisor_hostname'] not in failed_bm_ids]
+
             hvs = self.nova.hypervisors.list()
 
             failed_hv_ids = [str(hv.id) for hv in hvs
                              if hv.state == 'down' or hv.status == 'disabled']
-            failed_hosts = [host for host in reservable_hosts
-                            if host['id'] in failed_hv_ids]
+            failed_hosts = failed_hosts + [host for host in reservable_nova_hosts
+                                           if host['id'] in failed_hv_ids]
 
             active_hv_ids = [str(hv.id) for hv in hvs
                              if hv.state == 'up' and hv.status == 'enabled']
-            recovered_hosts = [host for host in unreservable_hosts
-                               if host['id'] in active_hv_ids]
+            recovered_hosts = recovered_hosts + [host for host in unreservable_nova_hosts
```

Comment:
I'm not sure it's safe to say that unreservable nodes are recovered. It would mean that if we for some reason set reservable to False for a node, and a user still had that node, it would not get kicked out of their lease, which would be a bit confusing for operators. What is the reason for treating unreservable nodes as recovered?